### PR TITLE
Proposal to re-enable C++ tests

### PR DESCRIPTION
**Pull request by parno**
I'd like to suggest we re-enable the tests in `Test/c++` as part of the CI, assuming the CI run for this PR succeeds.  That would help catch issues like those fixed in #1233.

### CODE REVIEW

**Comment by parno** [comment]
In addition to normal CI flakiness, it looks like we're starting to run into differences in C++ compilers on Linux vs. Mac.  For example, Rustan added `-Wno-unknown-warning-option` in #1131 to deal with the Mac version of `g++` not recognizing `-Wno-deprecated-copy`, but that is now causing the Linux version of `g++` to report:
```
+cc1plus: warning: unrecognized command line option '-Wno-unknown-warning-option'
```
Similarly, the `g++` install on the Linux CI machine appears to print `nullptr` as `0`, while the version on the Mac machines appears to print it as `0x0`.  

Any thoughts on how to reconcile these differences?  The warnings could be handled by disabling warnings entirely, or by having Dafny try to do OS detection (which seems a bit ugly).  The printing differences are a bit trickier.  We could try to add special cases to the C++ harness, or perhaps we should have different `.expect` files for different OSes.  Or I could just change the test case to not use null values.

**Comment by RustanLeino** [comment]
There seem to be differences in how `null` is printed (`0` versus `0x0`). Can something be done about that?

**Comment by parno** [comment]
It appears that Mac's version is not standards-compliant.  I just pushed a workaround based on C++ template magic, thanks to @jaybosamiya.  This should ensure that both Linux and Mac print `nullptr` consistently.  

That still leaves the problem of the Linux and Mac compilers accepting different warning-related options.  As mentioned above, we could disable compiler warnings entirely (and rely on people who update the compiler running with warnings on), or by having Dafny try to do OS detection (which seems a bit ugly).  @RustanLeino thoughts?

File: Binaries/DafnyRuntime.h [hunk 1]
```diff
@@ -73,6 +73,31 @@ class IntegerRange {
    iterator end_;
 };
 
+// Workaround the fact that Apple's clang and g++ print nullptr as 0x0,
+// while Linux's g++ prints it as 0
```
Comment (robin-aws) [inline Binaries/DafnyRuntime.h:77]: Nit-picking for the record: both `0` and `0x0` would be wrong anyway. :) The fact that Dafny null values are materialized as `nullptr` in C++ is pure implementation detail, and although Dafny doesn't currently specify enough about how arbitrary values are string-ified by `print` statements, it should be the same for each target language.

Comment (parno) [reply Binaries/DafnyRuntime.h:77]: Agreed that ideally all the backends should agree on print formatting.

Comment (RustanLeino) [reply Binaries/DafnyRuntime.h:77]: It's ideal if printing can be consistent everywhere and, more to the point, print things that make sense Dafny-wise. However, in the end, the `print` statement is for less demanding printing tasks, where the exact formatting probably doesn't matter. A demanding user would want to have more control over formatting, and would want to print not just to stdout. So, we may not always meet this ideal.

**Comment by RustanLeino** [comment]
I think we should disable warnings when possible, since they would appear to a Dafny user as if something went wrong during compilation (which is not the case here). So, of the various solutions, I thinking that having Dafny check which OS it's running on and choosing the flags according is the best.

**Comment by parno** [comment]
Okay, I added some OS-based checks to control which warning-related options we pass to `g++`, which seems to have satisfied the CI.  

On a side note, I notice that GitHub is saying:
> The ubuntu-16.04 environment is deprecated and will be removed on September 20, 2021. Migrate to ubuntu-latest instead. For more details see https://github.com/actions/virtual-environments/issues/3287

so we may need to make some adjustments when we change to a newer version of Ubuntu for testing.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**